### PR TITLE
Update Calculator keyboard for greater consistency and extensibility

### LINKED ISSUE #428: Update Calculator keyboard for greater consistency and extensibility
> This feature is in pre-production. Contribute to the in-progress spec **[here](https://github.com/Microsoft/calculator-specs/tree/keyboardrefresh/active/KeyboardRefresh)**.

**Problem Statement**
The Calculator keyboard requires some changes to meet the demands of upcoming features, including graphing mode and always-on-top mode. Furthermore, the current standard keyboard could be better at accommodating scaled screen sizes, which improves overall accessibility of the app. There are a number of other problems this spec seeks to address, including reducing confusion over the up arrow used as the shift key and reducing the number of keyboard layout states.

**Evidence or User Insights**
The current keyboard is not scalable to support new minimum size requirements for CompactOverlay mode, nor is it extensible to support additional functionality in graphing calculator.

**Proposal**
Update they keyboard app-wide for increased consistency and greater extensibility to prepare for upcoming feature improvements.

**Goals**
* No functionality takebacks in existing calculator modes
* Opportunistically add missing functionality across modes
* Increase clarity around shifted operations

**Non-Goals**
* Changes to display option buttons are out of scope (e.g., BYTE/WORD mode in programmer calculator), unless the button is better suited to be included in the keyboard

**Low-Fidelity Concept**
image

### PR DESCRIPTION
**Pull request by joseartrivera**
## Fixes #428

Spec for this change can be found here:
[Update Calculator keyboard for greater consistency and extensibility](https://github.com/microsoft/calculator-specs/blob/master/active/KeyboardRefresh/README.md)

### Description of the changes:
This PR includes the UI changes for the keyboard refresh spec.
- Added corresponding UI buttons for new calc functions added in #662
- Rearranged some of the buttons in the Standard, Scientific, and Programmer modes
- Added OperatorPanelButton that displays a flyout when clicked, storing buttons that are closely related
- C and CE button are conditionally visible based on if there is input that can be cleared
- Calc Buttons can be navigated with arrow keys
- Visual State layouts tweaked and simplified per the spec linked above
- Enable negate buttons in some conversion modes per the spec linked above

### How changes were validated:
- Manual testing

### CODE REVIEW

File: src/Calculator/Views/CalculatorScientificOperators.xaml [hunk 1]
```diff
@@ -233,383 +240,917 @@
...
+                            <!--
+                                Because of the way we handle keyboard shortcuts, the IsEnabled property must be bound to a button outside the flyout.
+                                This is because the content from within the Flyout do not inherit the IsEnabled property of the flyout parent,
+                                causing the shortcut keys to be used when the control would normally be disabled.
+                            -->
+                            <ToggleButton x:Name="TrigShiftButton"
+                                          x:Uid="trigShiftButton"
+                                          Style="{StaticResource CaptionToggleEmphasizedButtonStyle}"
+                                          FontFamily="{StaticResource CalculatorFontFamily}"
+                                          AutomationProperties.AutomationId="shiftButton"
```
Comment (joseartrivera) [inline src/Calculator/Views/CalculatorScientificOperators.xaml]: TODO: Need to fix AutomationId in this file and others, also need to add AutomationIds to OperatorPanelButtons

File: src/Calculator/Controls/OperatorPanelListView.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+{
+    m_scrollLeft->Visibility = ::Visibility::Collapsed;
+    m_scrollRight->Visibility = ::Visibility::Collapsed;
+    m_isPointerEntered = false;
+}
+
+void OperatorPanelListView::ScrollLeft()
+{
+        double offset = m_scrollViewer->HorizontalOffset - (scrollRatio * m_scrollViewer->ViewportWidth);
+        m_scrollViewer->ChangeView(offset, nullptr, nullptr);
```
Comment (sanderl) [inline src/Calculator/Controls/OperatorPanelListView.cpp]: nit: spacing and below

File: src/Calculator/Resources/en-US/Resources.resw [hunk 3]
```diff
@@ -3425,7 +3533,319 @@
...
+    <comment>{Locked}This is the character that should trigger this button. Note that it is a character and not a key, so it does not come from the Windows::System::VirtualKey enum.</comment>
+  </data>
+  <data name="nandButton.[using:Windows.UI.Xaml.Automation]AutomationProperties.Name" xml:space="preserve">
+    <value>Nand</value>
+    <comment>Screen reader prompt for the Calculator button nand in the scientific flyout keypad</comment>
+  </data>
+  <data name="nandButton.[using:CalculatorApp.Common]KeyboardShortcutManager.Character" xml:space="preserve">
+    <value>.</value>
+    <comment>{Locked}This is the character that should trigger this button. Note that it is a character and not a key, so it does not come from the Windows::System::VirtualKey enum.</comment>
+  </data>
```
Comment (sanderl) [inline src/Calculator/Resources/en-US/Resources.resw]: does this conflict with decimal?

File: src/Calculator/Views/CalculatorProgrammerRadixOperators.xaml.cpp [hunk 4]
```diff
@@ -29,32 +29,92 @@ CalculatorProgrammerRadixOperators::CalculatorProgrammerRadixOperators()
...
+        RolCarryButton->Visibility = ::Visibility::Visible;
+        RorCarryButton->Visibility = ::Visibility::Visible;
+        RolCarryButton->IsEnabled = true;
+        RorCarryButton->IsEnabled = true;
+    }
+
+    this->BitShiftFlyout->Hide();
+}
+
+void CalculatorProgrammerRadixOperators::collapseBitshiftButtons()
```
Comment (sanderl) [inline src/Calculator/Views/CalculatorProgrammerRadixOperators.xaml.cpp]: >c [](start = 41, length = 1)

capitalization: CollapseBitshiftButtons()

File: src/Calculator/Views/CalculatorProgrammerRadixOperators.xaml.h [hunk 5]
```diff
@@ -30,8 +30,12 @@ namespace CalculatorApp
 
         DEPENDENCY_PROPERTY_OWNER(CalculatorProgrammerRadixOperators);
 
+        void checkDefaultBitShift();
+
     private:
-        void Shift_Clicked(Platform::Object ^ sender, Windows::UI::Xaml::RoutedEventArgs ^ e);
+        void bitshiftFlyout_Checked(Platform::Object ^ sender, Windows::UI::Xaml::RoutedEventArgs ^ e);
```
Comment (sanderl) [inline src/Calculator/Views/CalculatorProgrammerRadixOperators.xaml.h]: Fix capitalization on these methods.

**Comment by sanderl** [comment]
>                       FontSize="20"

Should the font size always be 20?

---
Refers to: src/Calculator/Views/CalculatorScientificOperators.xaml:878 in 8c260fa. [](commit_id = 8c260fa4b749ca822b0d150f66005abf01ceca1f, deletion_comment = False)

**Comment by sanderl** [comment]
>                                    Content="e"/>

Should the font sizes be constant here too?

---
Refers to: src/Calculator/Views/CalculatorScientificOperators.xaml:903 in 8c260fa. [](commit_id = 8c260fa4b749ca822b0d150f66005abf01ceca1f, deletion_comment = False)

**Comment by sanderl** [comment]
>                                        Content="&#xE94F;"/>

constant font sizes

---
Refers to: src/Calculator/Views/CalculatorScientificOperators.xaml:943 in 8c260fa. [](commit_id = 8c260fa4b749ca822b0d150f66005abf01ceca1f, deletion_comment = False)

**Comment by sanderl** [comment]
>                                    FontSize="19"

constant font size for this button and the next 2 buttons as well

---
Refers to: src/Calculator/Views/CalculatorScientificOperators.xaml:1122 in 8c260fa. [](commit_id = 8c260fa4b749ca822b0d150f66005abf01ceca1f, deletion_comment = False)

File: src/Calculator/Views/CalculatorScientificOperators.xaml.h [hunk 6]
```diff
@@ -29,21 +29,19 @@ namespace CalculatorApp
...
 
     private:
-        void ShortLayout_Completed(_In_ Platform::Object ^ sender, _In_ Platform::Object ^ e);
-        void WideLayout_Completed(_In_ Platform::Object ^ sender, _In_ Platform::Object ^ e);
         void OnIsErrorVisualStatePropertyChanged(bool oldValue, bool newValue);
         void shiftButton_Check(_In_ Platform::Object ^ sender, _In_ Windows::UI::Xaml::RoutedEventArgs ^ e);
+        void shiftButton_Uncheck(_In_ Platform::Object ^ sender, _In_ Windows::UI::Xaml::RoutedEventArgs ^ e);
+        void trigFlyoutShift_Toggle(_In_ Platform::Object ^ sender, _In_ Windows::UI::Xaml::RoutedEventArgs ^ e);
+        void trigFlyoutHyp_Toggle(_In_ Platform::Object ^ sender, _In_ Windows::UI::Xaml::RoutedEventArgs ^ e);
+        void flyoutButton_Clicked(_In_ Platform::Object ^ sender, _In_ Windows::UI::Xaml::RoutedEventArgs ^ e);
```
Comment (sanderl) [inline src/Calculator/Views/CalculatorScientificOperators.xaml.h]: should these be capitals?

File: src/Calculator/Views/CalculatorStandardOperators.xaml [hunk 7]
```diff
@@ -350,45 +233,28 @@
...
-            <controls:CalculatorButton x:Name="SquareRootButton"
-                                       x:Uid="squareRootButton"
-                                       Grid.Column="1"
+            <controls:CalculatorButton x:Name="InvertButton"
+                                       x:Uid="invertButton"
                                        Style="{StaticResource SymbolOperatorButtonStyle}"
-                                       AutomationProperties.AutomationId="squareRootButton"
-                                       ButtonId="Sqrt"
-                                       Content="&#xF899;"/>
+                                       FontSize="18"
```
Comment (sanderl) [inline src/Calculator/Views/CalculatorStandardOperators.xaml]: >FontSize="18" [](start = 39, length = 13)

Should this always be 18?

File: src/CalculatorUITests/StandardModeFunctionalTests.cs [hunk 8]
```diff
@@ -242,13 +234,6 @@ public void KeyboardInput_SquareRoot()
             Assert.AreEqual("10", page.GetCalculatorResultText());
         }
 
-        [TestMethod]
-        public void KeyboardInput_Cubed()
-        {
-            page.Header.SendKeys("3#");
-            Assert.AreEqual("27", page.GetCalculatorResultText());
-        }
```
Comment (sanderl) [inline src/CalculatorUITests/StandardModeFunctionalTests.cs]: Is this test not valid any more or does it just need to be updated?

File: src/Calculator/Resources/en-US/Resources.resw (hunk 3 above)
Comment (joseartrivera) [reply src/Calculator/Resources/en-US/Resources.resw]: No, decimal is disabled in programmer mode

File: src/CalculatorUITests/StandardModeFunctionalTests.cs (hunk 8 above)
Comment (joseartrivera) [reply src/CalculatorUITests/StandardModeFunctionalTests.cs]: Not valid, button no longer exists in standard mode

File: src/Calculator/Views/CalculatorStandardOperators.xaml (hunk 7 above)
Comment (joseartrivera) [reply src/Calculator/Views/CalculatorStandardOperators.xaml]: These get overwritten by the visual sates. Same for other comments regarding fonts.

**Comment by rudyhuyn** [comment]
I'm a little late but congratulations on this PR, it looks a lot better! 🎉🍾

**Pull request merged**